h2o-algos/src/main/java/hex/gam/GAM.java

```diff
@@ -70,15 +81,47 @@ public GAM(GAMModel.GAMParameters parms, Key<GAMModel> key) {
     super(parms, key);
     init(false);
   }
-  
-  double[][] _knots;
 
-  /***
-   * This method will look at the keys of knots stored in _parms._knot_ids and copy them over to double[][]
-   * array.
-   *
-   * @return double[][] array containing the knots specified by users
-   */
+  // cross validation can be used to choose the best alpha/lambda values among a whole collection of alpha
+  // and lambda values.  Future hyperparameters can be added for cross-validation to choose as well.
+  @Override
+  public void computeCrossValidation() {
+    if (error_count() > 0) {
+      throw H2OModelBuilderIllegalArgumentException.makeFromBuilder(GAM.this);
+    }
+    _cvOn = true;  // set cross-validation on to be true
+    _validKeys = new IcedHashSet<>();
+    super.computeCrossValidation();
+  }
+
+  // find the best alpha/lambda values used to build the main model moving forward by looking at the devianceValid
+  @Override
+  public void cv_computeAndSetOptimalParameters(ModelBuilder[] cvModelBuilders) {
+    double deviance_valid = Double.POSITIVE_INFINITY;
+    double best_alpha = 0;
+    double best_lambda = 0;
+    for (int i = 0; i < cvModelBuilders.length; ++i) {  // run cv for each lambda value
+      GAMModel g = (GAMModel) cvModelBuilders[i].dest().get();
+      if (g._output._devianceValid < deviance_valid) {
+        best_alpha= g._output._best_alpha;
+        best_lambda = g._output._best_lambda;
+      }
+      
+      g.write_lock(_job);
+      g.update(_job);
+      g.unlock(_job);
+    }
+    _cv_alpha = new double[]{best_alpha};
+    _cv_lambda = new double[]{best_lambda};
+    _doInit = false;  // for main model, disable the init()
```

Comment:
this sounds suspicious and/or redundant - the main model already knows it is the main model